utils/pdetach.1
.\" pdetach - execute detached processes.
.\" Copyright (C) 2009-2014 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH args "1" "January 2010" "GNU uCommon" "GNU Telephony"
.SH NAME
pdetach \- execute detached processes
.SH SYNOPSIS
.B args
.RI [ stdio options ]
.B command
.RI [ args... ]
.br
.SH DESCRIPTION
This command is used to execute processes that are fully detached from the
shell and the user's controlling terminal.  Detached processes will not exit
if the user logs out.  The stdio options can be manipulated to attach the
process to another device or file.
.SH OPTIONS
.TP
.BI \-\-error= path
Set stderr for the detached process.
.TP
.BI \-\-input= path
Set stdin for the detached process.
.TP
.BI \-\-output= path
Set stdout for the detached process.
.SH AUTHOR
.B pdetach
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to bug-commoncpp@gnu.org or bugs@gnutelephony.org.
.SH COPYRIGHT
Copyright \(co 2009-2014 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.




NEWS
uCommon Library NEWS -- history of visible user changes. Updated 2008-04-15.

uCommon C++ 0.4 codebase completed and ready for release with the introduction
of the GNU GPL V3.  This is a complete rewrite of UCommon as a C++ library for
threading and design patterns suitable for deeply embedded targets.

uCommon C++ 0.7 codebase now has core documentation and basic library theory in
header files.  This documentation can be generated with doxygen or by using
"make doxy".

uCommon C++ 1.0 release has been simplified, fully documented, and modified in
specific ways to simplify the migration and merge with GNU Common C++ to form
GNU Common C++ 2.0.

uCommon C++ 1.5 release begins the process of migrating GNU Common C++ into the
core GNU uCommon C++ codebase. Some GNU Common C++ classes, including tcp
streaming, xml parser, and persist, have been revised and integrated for the
1.5.0 release itself.  Other GNU Common C++ classes will be integrated as
requested and needed.  The 1.5/1.x series will actually be the last UCommon
release sets, as GNU uCommon C++ 2.0 will instead be released as GNU Common C++
2 2.0.

uCommon C++ 2.0 released under the GNU Lesser General Public License, Version 3
or later.

uCommon C++ 2.1 is considered a refactored abi release offering greater
clarity, consistency of use, and additional features that were migrated from
Common C++. The 2.1 release series may continue to introduce additional
classes.

ucommon 2.2 merges other ucommon related libraries to create a single ucommon
framework for application development.



utils/pdetach.cpp
// Copyright (C) 2010-2014 David Sugar, Tycho Softworks.
//
// This file is part of GNU uCommon C++.
//
// GNU uCommon C++ is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU uCommon C++ is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with GNU uCommon C++.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/ucommon.h>

using namespace UCOMMON_NAMESPACE;

static shell::flagopt helpflag('h',"--help",    _TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::stringopt error('e', "--error", _TEXT("stderr path to use"), "filename");
static shell::stringopt input('i', "--input", _TEXT("stdin path to use"), "filename");
static shell::stringopt output('o', "--output", _TEXT("stdout path to use"), "filename");

PROGRAM_MAIN(argc, argv)
{
    fd_t stdio[3] = {INVALID_HANDLE_VALUE, INVALID_HANDLE_VALUE, INVALID_HANDLE_VALUE};
    const char *argv0;

    shell::bind("pdetach");
    shell args(argc, argv);

    if(is(helpflag) || is(althelp)) {
        printf("%s\n", _TEXT("Usage: pdetach [stdio options] command [arguments...]"));
        printf("%s\n\n", _TEXT("Create detached process"));
        printf("%s\n", _TEXT("Options:"));
        shell::help();
        printf("\n%s\n", _TEXT("Report bugs to dyfet@gnu.org"));
        PROGRAM_EXIT(0);
    }

    if(!args())
        shell::errexit(10, "*** pdetach %s", _TEXT("no command specified"));

    argv = args.argv();
    argv0 = *argv;

    if(is(input)) {
        stdio[0] = fsys::input(*input);
        if(stdio[0] == INVALID_HANDLE_VALUE)
            shell::errexit(1, "*** pdetach: %s: %s",
                _TEXT("cannot access"), *input);
    }

    if(is(output)) {
        stdio[1] = fsys::output(*output);
        if(stdio[1] == INVALID_HANDLE_VALUE)
            shell::errexit(2, "*** pdetach: %s: %s",
                _TEXT("cannot access"), *output);
    }

    if(is(error)) {
        stdio[2] = fsys::output(*error);
        if(stdio[2] == INVALID_HANDLE_VALUE)
            shell::errexit(3, "*** pdetach: %s: %s",
                _TEXT("cannot access"), *error);
    }

    if(shell::detach(argv0, argv, NULL, stdio))
        shell::errexit(-1, "*** pdetach: %s: %s",
            argv0, _TEXT("failed to execute"));

    PROGRAM_EXIT(0);
}

